src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2841,8 +2846,73 @@ private void CheckForStructBadInitializers(MembersAndInitializersBuilder builder
             }
         }
 
+        private void AddSynthesizedRecordMembersIfNecessary(ArrayBuilder<Symbol> members, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(declaration.Kind == DeclarationKind.Class || declaration.Kind == DeclarationKind.Struct);
+
+            ParameterListSyntax? paramList = null;
+            foreach (SingleTypeDeclaration decl in declaration.Declarations)
+            {
+                var syntaxNode = decl.SyntaxReference.GetSyntax();
+                ParameterListSyntax? curParamList = null;
+                switch (declaration.Kind)
+                {
+                    case DeclarationKind.Class:
+                        var classDecl = (ClassDeclarationSyntax)syntaxNode;
+                        curParamList = classDecl.ParameterList;
+                        break;
+                    case DeclarationKind.Struct:
+                        var structDecl = (StructDeclarationSyntax)syntaxNode;
+                        curParamList = structDecl.ParameterList;
+                        break;
+                }
+
+                if (paramList is null)
+                {
+                    paramList = curParamList;
+                }
+                else
+                {
+                    // PROTOTYPE: Add error for multiple parameter lists
+                }
+            }
+
+            if (paramList is null)
+            {
+                if (_declModifiers.HasFlag(DeclarationModifiers.Data))
+                {
+                    diagnostics.Add(ErrorCode.ERR_BadRecordDeclaration, declaration.NameLocations[0]);
+                }
+                return;
+            }
+
+            if (!_declModifiers.HasFlag(DeclarationModifiers.Data))
+            {
+                diagnostics.Add(ErrorCode.ERR_BadRecordDeclaration, paramList.Location);
+            }
+
+            BinderFactory binderFactory = this.DeclaringCompilation.GetBinderFactory(paramList.SyntaxTree);
+            var binder = binderFactory.GetBinder(paramList);
+
+            var memberSignatures = new HashSet<Symbol>(members, MemberSignatureComparer.DuplicateSourceComparer);
+
+            var ctor = new SynthesizedRecordConstructor(this, binder, paramList, diagnostics);
+            if (!memberSignatures.Contains(ctor))
```

Comment:
Rather than constructing a `HashSet<>`, would it be more efficient to just call `members.Contains(ctor)`?

Reply:
But I don't want to know if it has the exact symbol, I want to know if it has any symbol with a conflicting signature.

Reply:
`members.Contains(ctor, MemberSignatureComparer.DuplicateSourceComparer)`

---
In reply to: [367670582](https://github.com/dotnet/roslyn/pull/40625#discussion_r367670582) [](ancestors = 367670582)

Reply:
Right, a linear walk. But this is just the ctor. I'm also gonna need to query for the other methods, like the generated properties, equals, hashcode, etc.